maitake/src/wait/cell.rs

```diff
@@ -306,45 +399,63 @@ pub(crate) mod test_util {
 #[cfg(all(loom, test))]
 mod loom {
     use super::*;
-    use crate::loom::{future, thread};
-
-    const NUM_NOTIFY: usize = 2;
+    use crate::loom::{future, sync::Arc, thread};
+    use futures::{select_biased, FutureExt};
 
     #[test]
-    fn basic_latch() {
+    fn basic() {
         crate::loom::model(|| {
-            let chan = test_util::Chan::new(NUM_NOTIFY);
-
-            for _ in 0..NUM_NOTIFY {
-                let chan = chan.clone();
-
-                thread::spawn(move || chan.notify());
-            }
-
-            future::block_on(chan.wait());
-        });
-    }
+            let wake = Arc::new(WaitCell::new());
 
-    #[test]
-    fn close() {
-        crate::loom::model(|| {
-            let chan = test_util::Chan::new(NUM_NOTIFY);
+            let waiter = wake.clone();
+            let closer = wake.clone();
 
-            thread::spawn({
-                let chan = chan.clone();
-                move || {
-                    chan.notify();
-                }
+            thread::spawn(move || {
+                info!("waking");
+                waiter.wake();
+                info!("woke'd");
+            });
+            thread::spawn(move || {
+                info!("closing");
+                closer.close();
+                info!("closed");
             });
 
-            thread::spawn({
-                let chan = chan.clone();
-                move || {
-                    chan.close();
+            info!("waiting");
+            let _ = future::block_on(async move {
+                // Unfortunately, SOMETIMES loom just wants to preempt the 'waker' task
+                // JUST as it gains access to the lock, but BEFORE actually waking the pending task,
+                // never yielding back to the waker task, which then just essentially leaves it stuck
+                // in the NOTIFYING state.
```

Comment:
this...feels like kind of a serious bug, to me? i left some comments on the `register_wait` function about what i _think_ we should do here, but it's possible i've forgotten some of the context...